Implement a boxplot

Hoid opened the pull request:
This PR addresses the following things:
1. Creates new chart subclasses called `LineChart` and `BoxplotChart` and makes the `Chart` class an abstract base class
2. Changes the structure of the callbacks and where the chart is being used in `visualization_editor.py` so that there's a single `dcc.Graph` object created with the id of `graph` and the figure inside it is the only thing that ever gets updated by the callbacks, not the `Graph` object itself
3. Fixes the click-in-whitespace bug (likely because of number 2)
4. Refactors some littler things
5. Fixes a smaller bug or two

Image of the current boxplot is below

<img width="1220" height="769" alt="image" src="https://github.com/user-attachments/assets/f7b8a512-1b97-42be-9382-f6cbd89fc707" />
<img width="1161" height="572" alt="image" src="https://github.com/user-attachments/assets/f1d84222-117e-47ba-999c-4d40fcfe16be" />
<img width="1216" height="712" alt="image" src="https://github.com/user-attachments/assets/e3f24bf7-69c7-4e5d-9830-7b952db1ea54" />

Hoid committed:
Allow passing in plot type to chart controls

Hoid committed:
Merge branch 'dev' of https://github.com/RENCI/public-health-app into feature/boxplot

Hoid committed:
Add chart-controls-store and use it

Hoid committed:
Fix callbacks and use graph figure

Hoid committed:
Make the chart controls show up for my slightly smaller screen

Hoid commented:
@mbwatson Looks like in the `visualization_editor.py` file I changed the callback for `chart-extent-store` and never added it back. But can you take a look at what I have and let me know if it should work this way without `chart-extent-store` since every time the graph's `relayout` changes I'm now just recreating the chart object? (In the future maybe we can optimize that new callback so that it only changes the existing chart object instead of creating a new one, but that's for another time)

Hoid commented:
There are a couple things that will be improved about the boxplot before I merge:
1. The boxplot chart currently overlays the second and fourth subplots over each other instead of being separate things. I'll look at this tomorrow morning.
2. The boxplots are currently vertical instead of horizontal. This should be an easy fix that I can do tomorrow.
3. The boxplots show up in the viewer page but not in the explorer page; the boxplots get replaced with line charts for the same data.
4. I'm going to move some things from the chart.py file into their own files.

davidborland commented:
@Hoid How do I create a boxplot in the interface?

Hoid commented:
Currently it's manually set in the template files. There are two insights in round 19: the first is a line plot and the second is a boxplot

davidborland commented:
Looking good so far. We will want to think about how to create a boxplot from scratch if there is no current boxplot insight to modify.

Hoid committed:
Split charts and stuff into their own files and implement certainty interval colors

Hoid committed:
Interim commit (broken)

Hoid committed:
Merge branch 'dev' of https://github.com/RENCI/public-health-app into feature/boxplot

Hoid committed:
Merge from dev

Hoid committed:
Suppress failures in callbacks for now

Hoid committed:
Fix dark theme and add a hack for getting the plot type to update on page load

Hoid committed:
Clean up some initial control values stuff

Hoid committed:
Fix a bug with visualization_editor

Hoid committed:
Clean up boxplot

Hoid committed:
Set unified axes for boxplots

Hoid committed:
Fix annotations and remove a couple prints

Hoid committed:
Disable certain controls if you're looking at a boxplot

Hoid commented:
This might be ready for review now! It's working pretty well, I just don't yet have the ability to create a boxplot from scratch yet, it has to be set as one in the yaml. I figure this might be fine for now and we can add the ability to change the plot type to a boxplot later, just lemme know what y'all think.

Hoid committed:
Fix some comments

Hoid committed:
Merge branch 'dev' of https://github.com/RENCI/public-health-app into feature/boxplot

davidborland reviewed (changes requested):
I think this is looking good overall. One thing I think still needs a little work is the chart annotation. It would be good if the box plot matched he line plot as much as possible:

<img width="1180" height="842" alt="image" src="https://github.com/user-attachments/assets/2ce6602d-0b2b-4b8f-b011-9bdba5f1142c" />

So for the box plot we would want:
* Title from currently selected Target
* Subtitle from pathogen/location/age group
* X-axis title from currently selected Target

Might also be good to include the date range in the title, since that is not represented in the chart, e.g.: 
* Incident Hospitalization distribution, 1/1/2020 - 2/2/2020

Hoid committed:
Change certainty percent -> uncertainty interval everywhere

Hoid committed:
Add chart title/subtitle to boxplot and fix chart title for line chart

Hoid committed:
Fix a bug in the insight viewer

Hoid committed:
Fix a console error about css styles not getting loaded

Hoid committed:
Implement date ranges for chart based on data

Hoid commented:
> I think this is looking good overall. One thing I think still needs a little work is the chart annotation. It would be good if the box plot matched he line plot as much as possible:
> 
> So for the box plot we would want:
> 
>     * Title from currently selected Target
> 
>     * Subtitle from pathogen/location/age group
> 
>     * X-axis title from currently selected Target
> 
> 
> Might also be good to include the date range in the title, since that is not represented in the chart, e.g.:
> 
>     * Incident Hospitalization distribution, 1/1/2020 - 2/2/2020

Okay I did the following:
* Added a title and subtitle that seem relevant to the boxplot
* Updated the line chart subtitle to have the uncertainty interval
* Changed "certainty percent" to "uncertainty interval" everywhere
* Fixed a bug in the insight viewer where it would complain about having no annotations
* Fixed a console error about css styles not getting loaded

I feel like adding the target to the X axis (or whichever is showing the data) might be superfluous since it's also in the title of the chart. This wouldn't be a big deal for one boxplot but once you get to two or more I feel like it'd be better to just keep the target in the title, or in the case of multiple columns just put the target at the top of the column instead of on every X axis. What do you think?

Also, each scenario doesn't have metadata for the start and end dates of each scenario, so I'm inferring the start date from the scenario name and pulling the end date as the max value of `target_end_date` in the data for that scenario and displaying the latest end date for all scenarios. That info is now in the chart subtitle. The start date is currently inaccurate because the scenario names we are using aren't for the right start dates, but I can fix that in a future PR if that's okay. For now it paints a general picture for the user.

davidborland commented:
Some more minor tweaks:
* Remove "forecast" and "(boxplot)" from boxplot title, e.g. "Incident Hospitalization distribution". The term "forecast" has a very specific meaning in the simulation world that implies a level of precision not provided here. We could potentially use "projection". If so I would add to both the boxplot and line plot, but I don't think it is needed. I think I would probably remove "(by scenario)" from the line plot as well, so we would have: "{Target} distribution" and "{Target} over time" for the titles.
* I like the idea of having the X-axis title per column, but maybe put at the bottom? Not sure how easy that is if the column/no-column is determined by a responsive layout though.
* Can't you just get the first and last `target_end_date` and use that for the date range for the box plots?
* Conversely, I'm not sure we should put the date in the subtitle for the line plot, as that is already shown via the x-axis, and the data shown could change depending on the zoom.

Hoid commented:
>     * Can't you just get the first and last `target_end_date` and use that for the date range for the box plots?
> 
>     * Conversely, I'm not sure we should put the date in the subtitle for the line plot, as that is already shown via the x-axis, and the data shown could change depending on the zoom.

I guess I could use the target_end_date for both, I guess I just assumed that the target_end_date would always be the latest date of data for the scenario but that can't be right because we're using it as the time series axis... so yea I can do that.

For the subtitle, do you think then that we should _just_ use the time interval in the boxplot, instead?

Hoid committed:
Fix a bug where a markdown callback would attempt to update before rendering

Hoid committed:
Just remove some unnecessary whitespace from an insight yaml

Hoid committed:
Remove time interval from line chart subtitle and fix a bug

Hoid commented:
I basically ended up doing what I said in the last comment: only show the "During: ..." time interval on line charts. I also started using the earliest target_end_date for the start date in the time interval and fixed a random bug with local storage.

davidborland commented:
I think this looks good. Let's discuss the x-axis title in the 1:30 meeting, and then I will approve. We should create a new issue to capture the ability to create a box plot from the interface though.

Hoid commented:
Finished adding the X axis to the bottom plot. Here's a screenshot:
<img width="671" height="600" alt="image" src="https://github.com/user-attachments/assets/9951f59e-38f5-4825-9d89-4bc071216c8f" />

Hoid committed:
Add x axis label for the target, but only on bottom plot

davidborland reviewed (approved):
Looks good!
